packages/pg/lib/client.js

```diff
@@ -156,23 +312,366 @@ class Client extends EventEmitter {
     })
   }
 
-  connect(callback) {
+  attachErrorListenerOnClientConnection(client) {
+    client.on('error', () => {
+      if (Client.hostServerInfo.has(client.host)) {
+        Client.failedHosts.set(client.host, Client.hostServerInfo.get(client.host))
+        Client.connectionMap.delete(client.host)
+        Client.hostServerInfo.delete(client.host)
+      }
+      Client.controlClient = undefined
+    })
+  }
+
+  async getConnection() {
+    let currConnectionString = this.connectionString
+    var client = new Client(currConnectionString)
+    this.attachErrorListenerOnClientConnection(client)
+    let lookup = util.promisify(dns.lookup)
+    let addresses = [{ address: client.host }]
+    await lookup(client.host, { family: 0, all: true }).then((res) => {
+      addresses = res
+    })
+    client.host = addresses[0].address // If both resolved then - IPv6 else IPv4
+    client.loadBalance = false
+    client.connectionParameters.loadBalance = false
+    client.topologyKeys = ''
+    client.connectionParameters.topologyKeys = ''
+    if (Client.failedHosts.has(client.host)) {
+      let upHostsList = Client.hostServerInfo.keys()
+      let upHost = upHostsList.next()
+      let hostIsUp = false
+      while (upHost !== undefined && !hostIsUp) {
```

Comment:
This while loop is similar to the one in else block below. Can be put into a method which can be used at both these places for brevity. Something like
```
  iterateHostList(client) {
    let upHostsList = Client.hostServerInfo.keys()
    let upHost = upHostsList.next()
    let hostIsUp = false

    while (upHost !== undefined && !hostIsUp) {
      client.host = upHost.value
      client.connectionParameters.host = client.host
      await client
        .nowConnect()
        .then((res) => {
          hostIsUp = true
        })
        .catch((err) => {
          client.connection =
            client.config.connection ||
            new Connection({
              stream: client.config.stream,
              ssl: client.connectionParameters.ssl,
              keepAlive: client.config.keepAlive || false,
              keepAliveInitialDelayMillis: client.config.keepAliveInitialDelayMillis || 0,
              encoding: client.connectionParameters.client_encoding || 'utf8',
            })
          client._connecting = false
          upHost = upHostsList.next()
        })
    }
  }
```